pkg/goclient/goclient.go

```diff
@@ -0,0 +1,177 @@
+*/
+func initFunc() (uds.CleanupFunc, error) {
+	hWR = uds.NewHandler()
+	hPod = host.NewHandler()
+	var response string
+
+	// init uds Handler for reading and writing
+	if err := hWR.Init(constants.Uds.PodPath, constants.Uds.Protocol, constants.Uds.MsgBufSize, constants.Uds.CtlBufSize, 0*time.Second, ""); err != nil {
+		return nil, fmt.Errorf("Library Error: Error Initialising UDS server: %v", err)
+	}
+
+	cleanupG, err := hWR.Dial()
+	if err != nil {
+		return cleanupG, fmt.Errorf("Library Error: UDS Dial error: %v", err)
+	}
+
+	hostname, err := hPod.Hostname()
+	if err != nil {
+		return nil, fmt.Errorf("Library Error: Failed to initialize host: %v", err)
+	}
+
+	if err = hWR.Write(constants.Uds.Handshake.RequestConnect+", "+hostname, -1); err != nil {
+		return nil, fmt.Errorf("Library Error: UDS Write error: %v", err)
+	}
+
+	if response, _, err = hWR.Read(); err != nil {
+		return nil, fmt.Errorf("Library Error: UDS Read error : %v", err)
+	}
+
+	if response == constants.Uds.Handshake.ResponseHostOk {
+		connected = true
+	}
+
+	return cleanupG, nil
+}
+
+/*
+GetVersion returns the version of our Handshake
+*/
+func GetVersion() (version string, cleaner uds.CleanupFunc, e error) {
+	if !connected {
+		cleanupG, e = initFunc()
+	}
+
+	if e != nil {
+		return constants.Uds.Handshake.ResponseBadRequest,
+			nil,
+			fmt.Errorf("Library Error: %v", e)
+	}
+
+	if e := hWR.Write(constants.Uds.Handshake.RequestVersion, -1); e != nil {
+		return constants.Uds.Handshake.ResponseBadRequest,
+			nil,
+			fmt.Errorf("Library Error: %v", e)
+	}
+
+	_, _, e = hWR.Read()
+	if e != nil {
+		return constants.Uds.Handshake.ResponseBadRequest,
+			nil,
+			fmt.Errorf("Library Error: %v", e)
+
+	}
+
+	version = constants.Uds.Handshake.Version
+
+	return version,
+		cleanupG,
+		nil
+}
+
+/*
+RequestXSKmapFD requires a list of device names and returns a map of the fds for each device, and a cleanup function to close the connection
+*/
+func RequestXSKmapFD(devNames []string) (map[string]int, uds.CleanupFunc, error) {
+	if !connected {
+		cleanup, err := initFunc()
+		if err != nil {
+			return nil,
+				nil,
+				fmt.Errorf("Library Error: %v", err)
+
+		}
+		cleanupG = cleanup
+	}
+
+	fds := make(map[string]int)
+
+	for _, dev := range devNames {
+		if err := hWR.Write(constants.Uds.Handshake.RequestFd+", "+dev, -1); err != nil {
+			return nil,
+				nil,
+				fmt.Errorf("Library Error: UDS Write error: %v", err)
+
+		}
+		response, fd, err := hWR.Read()
+		if err != nil {
+			return nil,
+				nil,
+				fmt.Errorf("Library Error: UDS Read error: %v", err)
+
+		}
+		if response == constants.Uds.Handshake.ResponseFdAck {
+			fds[dev] = fd
+		}
+	}
+
+	return fds, cleanupG, nil
+}
+
+/*
+RequestBusyPoll takes a timeout, budget and a fd to request the busypoll for a specific device, and returns an fd, response, cleanup function and error
+*/
+func RequestBusyPoll(busyTimeout, busyBudget, fd int) (int, string, uds.CleanupFunc, error) {
```

Comment:
I think you've made too much work for yourself here!

The code calling this doesn't expect anything to be returned (just the cleanup and error, as usual). There's nothing to be returned here really, once BP was configured and there were no errors, the code calling this should be happy.

So when you Read() at line 165, your code (the code that understands the handshake) should check the response. Was it a responseBusyPollAck or responseBusyPollAck , etc.

If you did get a /nak, then you should create and return a meaningful error.

The code calling this will understand that an error occurred, but the purpose of the library is we're trying to get to the point where it doesn't have to be aware of the handshake. We want to interpret the responses here, so the calling code doesn't have to.

So for example, you get a nak response, try build an error with the useful info. You received a response, so it's not a read/write/uds problem. You didn't receive a badrequest so you know your request was good and the device plugin understands your request. So it's something that happened while the DP was attempting to set busy poll. So something like "Device plugin error configuring busy poll" will be very helpful and will hint that the device plugin logs may contain more detail.

Reply:
You should also check the response above in your xsk function.
you're returning the FD, but if there was an error, you'll receive a /nak or similar and the fd will simply be zero I think (I'd need to check the code).

So there you could also check and interpret the response string, if it was anything other than "/ack" then you should return an error with a fd of zero.

Reply:
Quick question again, so should my xsk map fd function just take one device and return an fd for it?

Reply:
yea, that makes sense actually..